src/misc/messages.cpp

```diff
@@ -62,49 +83,89 @@ class Message {
 		if (rendered_msg.empty())
 			rendered_msg = convert_ansi_markup(markup_msg.c_str());
 		assert(rendered_msg.length());
-		return rendered_msg.c_str();
+
+		if (!is_utf8_msg)
+			return rendered_msg.c_str();
+
+		const uint16_t cp = UTF8_GetCodePage();
+		if (rendered_cp_msg[cp].empty()) {
+			UTF8_RenderForDos(rendered_msg, rendered_cp_msg[cp], cp);
+			assert(rendered_cp_msg[cp].length());
+		}
+		return rendered_cp_msg[cp].c_str();
 	}
 
-	void Set(const char *markup)
+	void Set(const char *markup, const bool is_utf8)
 	{
 		assert(markup);
-		markup_msg = markup;
+		markup_msg  = markup;
+		is_utf8_msg = is_utf8;
+
 		rendered_msg.clear();
+		markup_cp_msg.clear();
+		rendered_cp_msg.clear();
 	}
 };
 
 static std::unordered_map<std::string, Message> messages;
 static std::deque<std::string> messages_order;
 
-// Add the message if it doesn't exist yet
-void MSG_Add(const char *name, const char *markup_msg)
+static void msg_add(const char *name, const char *markup_msg, const bool is_utf8)
 {
-	const auto &pair = messages.try_emplace(name, markup_msg);
+	const auto &pair = messages.try_emplace(name, markup_msg, is_utf8);
 	if (pair.second) // if the insertion was successful
 		messages_order.emplace_back(name);
 }
 
+// Add the message if it doesn't exist yet
+void MSG_Add(const char *name, const char *markup_msg)
+{
+	msg_add(name, markup_msg, false);
+}
+
 // Replace existing or add if it doesn't exist
-void MSG_Replace(const char *name, const char *markup_msg)
+static void msg_replace(const char *name, const char *markup_msg, const bool is_utf8)
 {
 	auto it = messages.find(name);
 	if (it == messages.end())
-		MSG_Add(name, markup_msg);
+		msg_add(name, markup_msg, is_utf8);
 	else
-		it->second.Set(markup_msg);
+		it->second.Set(markup_msg, is_utf8);
 }
 
-static bool LoadMessageFile(const std_fs::path &filename)
+static bool load_message_file(const std_fs::path &filename_plain,
+                              const std_fs::path &filename_utf8)
 {
-	if (filename.empty())
+	if (filename_plain.empty() && filename_utf8.empty())
 		return false;
 
-	const auto filename_str = filename.string();
+	auto check_file = [](const std_fs::path &filename) {
+		if (filename.empty())
+			return false;
+		return (std_fs::status(filename).type() !=
+		        std_fs::file_type::not_found);
+	};
+
+	bool is_utf8 = false;
+
+	std::string filename_str;
+	if (check_file(filename_utf8)) {
+		filename_str = filename_utf8.string();
+		is_utf8      = true;
+	} else if (check_file(filename_plain))
+		filename_str = filename_plain.string();
+	else {
+		std::string out_str = "";
```

Comment:
With C++17, the `std::filesystem::path` has a `c_str()` member (letting us skip the `.string()` intermediary: https://en.cppreference.com/w/cpp/filesystem/path/native


`std::string filename_str` and `out_str` can be held as `std_fs::path`, and then the `.string()` skipped.

Reply:
Changed.

Reply:
Unfortunately, it seems we can't get rid of using `string()` in `fopen` call, MSVC does not like this:

`error C2664: 'FILE *fopen(const char *,const char *)': cannot convert argument 1 from 'const std::filesystem::path::value_type *' to 'const char *'`

Reply:
And there are MinGW warnings. We can't skip `string()`, some compilers don't like it.

Reply:
Oh, darn! And sorry for this. Comment wasting both our time; let's hope the path object is fully compliant in these compilers when we switch to c++20.